clients/drcachesim/drptpostprocessor/drptpostprocessor.c

```diff
@@ -0,0 +1,649 @@
+
+    /* The perf event sideband decoder configuration. */
+    struct pt_sb_pevent_config sb_pevent_conf;
+
+    /* The sideband session. */
+    struct pt_sb_session *sb_session;
+} dript_decoder_t;
+
+/* A collection of options. */
+typedef struct _dript_options_t {
+    /* Print statistics. */
+    uint32_t print_stats : 1;
+} dript_options_t;
+
+/* A collection of output. */
+typedef struct _dript_output_t {
+    /* The number of instructions. */
+    uint64_t instr_count;
+} dript_output_t;
+
+/* Diagnostic output for error event.
+ * It will be called when 'pt_insn_sync_forward' return error status or decoder end before
+ * reach pte_eos.
+ */
+static void
+diagnose_error(struct pt_insn_decoder *decoder IN, int errcode IN, const char *errtype IN,
+               uint64_t ip IN)
+{
+    int err;
+    uint64_t pos;
+
+    err = -pte_internal;
+    pos = 0ull;
+
+    err = pt_insn_get_offset(decoder, &pos);
+
+    if (err < 0) {
+        fprintf(stderr, "could not determine offset: %s\n", pt_errstr(pt_errcode(err)));
+        fprintf(stderr, "[?, %" PRIx64 ": %s: %s]\n", ip, errtype,
+                pt_errstr(pt_errcode(errcode)));
+    } else
+        fprintf(stderr, "[%" PRIx64 ", %" PRIx64 ": %s: %s]\n", pos, ip, errtype,
+                pt_errstr(pt_errcode(errcode)));
+}
+
+static void
+process_decode(dript_decoder_t *decoder IN, const dript_options_t *options IN,
+               dript_output_t *output OUT)
+{
+    for (;;) {
+        struct pt_insn insn;
+        int status;
+
+        memset(&insn, 0, sizeof(insn));
+
+        /* Sync decoder to first Packet Stream Boundary (PSB) packet, and then decode
+         * instructions. If there is no PSB packet, the decoder will be synced to the end
+         * of the trace. If error occurs, we will call diagnose_error to print the error
+         * information.
+         * What is PSB packets?
+         * Below answer is quoted from Intel® 64 and IA-32 Architectures Software
+         * Developer’s Manual 32.1.1.1 Packet Summary
+         * “Packet Stream Boundary (PSB) packets: PSB packets act as ‘heartbeats’ that are
+         * generated at regular intervals (e.g., every 4K trace packet bytes). These
+         * packets allow the packet decoder to find the packet boundaries within the
+         * output data stream; a PSB packet should be the first packet that a decoder
+         * looks for when beginning to decode a trace.”
+         */
+        status = pt_insn_sync_forward(decoder->instr_decoder);
+        if (status < 0) {
+            if (status == -pte_eos)
+                break;
+            diagnose_error(decoder->instr_decoder, status, "sync error", insn.ip);
+            break;
+        }
+
+        /* Decode instructions. */
+        for (;;) {
+            /* Handle next status and all pending perf events. */
+            int nextstatus = status;
+            int errcode = 0;
+            struct pt_image *image;
+            while (nextstatus & pts_event_pending) {
```

Comment:
There seem to be many nested loops. Can you add a comment over each one to describe what they are for?